### PR TITLE
Add deployment status and status date

### PR DESCRIPTION
**Pull request by mlqn**
## Description
Added deployment status and deployment status date

## Related Issue(s)
- https://github.com/Altinn/altinn-studio/issues/12119

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [x] All tests run green

### CODE REVIEW

**Comment by tjololo** [comment]
Not quite sure of the desired status for every given conditions as there are quite a few, but some observations when a new deployment is triggered:
| condition | observed status |
| -------------------------------------------------- | ------------------------------------ |
| Old pod running, new pod failing readinessProbe    | Available |
| Old pod running, new pod crashlooping                    | Available |
| Old pod gone, new pod running                                 | Completed |
| First deployment, pod failing readiness                     | Progressing |
| First deployment, pod failing readiness, progressDeadlineSeconds exeeded | Failed |
| Old and new pod crashlooping                                   | Progressing |
| Old pod running, new pod failing readiness, processDeadlineSeconds exeeded | Failed |

The failed status will first be presented once kubernetes gives up on progressing a deployment, that might not happen for quite some time (default: 10 minutes).
Right now the Available and Progressing types seems to be connected in the code I don't think that they are as tightly coupled in the "real world".
I might be mistaken, but I belive the Type Available does not take into account if there is an ongoing deployment, just that there is a pod that handles user requests.
Trying to report two status in one might be hard and lead to situations where the developer/deployer is given the wrong status.

So TL;DR; are we trying to show the status of the deployment (making a new version of an application receiving traffic), or are we trying to report the status of the kubernetes resources apps/v1.Deployment? I don't think its wise to try and report both these status in one.

**Comment by mlqn** [comment]
Thanks for the detailed review @tjololo! :)

Oh, I see your point, why mixing the Available and Progressing conditions is wrong. Yes, we are trying to show the status of the deployment, but I thought that the Available condition was part of the deployment prosess and that only the updated pods would affect this condition, not the old ones.
I removed the Available value from the status and also removed the availabilityPercentage as it does not seem to reflect the real number of updated and available pods.

Regarding the failed status, there don't seem to be any Failing statuses/reasons that I can use. There is a ReplicaFailure condition, but it doesn't exist when a pod is crashlooping.
But a solution I've tested that works well is using the `--atomic` flag with Helm, as recommended in this article: https://polarsquad.com/blog/check-your-helm-deployments.
This automatically triggers a rollback if the deployment is still failing after 5 minutes (Helm's timeout). This reduces the waiting time to 5 minutes, and most importantly, ensures a stable version is in the environment. KubernetesWrapper will then return the previous version and mark it as Completed.

**Comment by sduranc** [comment]
@mlqn nice that you're doing this!

- It's always good to check what the official tools are doing, in this case you can see that the Deployment controller is also [checking](https://github.com/kubernetes/kubernetes/blob/v1.25.0/pkg/controller/deployment/progress.go#L49) for status.replicas == status.UpdatedReplicas:
`	isCompleteDeployment := newStatus.Replicas == newStatus.UpdatedReplicas && currentCond != nil && currentCond.Reason == util.NewRSAvailableReason`
So I guess we can do the same?
- Since we're now using --atomic from Helm we can take a look at how do they [check](https://github.com/helm/helm/blob/main/pkg/kube/ready.go#L290) "readiness" for the resources:
 ```
 func (c *ReadyChecker) deploymentReady(rs *appsv1.ReplicaSet, dep *appsv1.Deployment) bool {
	// Verify the replicaset readiness
	if !c.replicaSetReady(rs) {
		return false
	}
	// Verify the generation observed by the deployment controller matches the spec generation
	if dep.Status.ObservedGeneration != dep.ObjectMeta.Generation {
		c.log("Deployment is not ready: %s/%s. observedGeneration (%d) does not match spec generation (%d).", dep.Namespace, dep.Name, dep.Status.ObservedGeneration, dep.ObjectMeta.Generation)
		return false
	}

	expectedReady := *dep.Spec.Replicas - deploymentutil.MaxUnavailable(*dep)
	if !(rs.Status.ReadyReplicas >= expectedReady) {
		c.log("Deployment is not ready: %s/%s. %d out of %d expected pods are ready", dep.Namespace, dep.Name, rs.Status.ReadyReplicas, expectedReady)
		return false
	}
	return true
}
```
I guess we could omit the Generation to make it simpler (they use it). The last part of the func, the `expectedReady`, seems to be almost the same as checking for the `Available` condition with reason `MinimumReplicasAvailable`. So could rely on that condition as well?

- I agree with @tjololo that we should only target Deployments here and not if the App (deploymeny/daemonset/etc) itself is healthy, as that's way harder to define. 
- Since we're using `--atomic`, this can end up in 2 states possibles for the deployment: Healthy or unhealthy. Heallthy are both Completed and Rollbacked. Unhealthy it's only temporary because it'll be rollbacked anyway, right? What I mean is that maybe you don't need to report on Failures because they're temporary — they last as much as the `progressDeadlineSeconds`.

Good discussion, this isn't a simple problem :)

**Comment by mlqn** [comment]
Thanks for your review @sduranc!

> - It's always good to check what the official tools are doing, in this case you can see that the Deployment controller is also [checking](https://github.com/kubernetes/kubernetes/blob/v1.25.0/pkg/controller/deployment/progress.go#L49) for status.replicas == status.UpdatedReplicas:
`	isCompleteDeployment := newStatus.Replicas == newStatus.UpdatedReplicas && currentCond != nil && currentCond.Reason == util.NewRSAvailableReason`
So I guess we can do the same?

Nice catch, I will fix it 👍 

> - The last part of the func, the `expectedReady`, seems to be almost the same as checking for the `Available` condition with reason `MinimumReplicasAvailable`. So could rely on that condition as well?

Yes, Helm will mark the deployment as successful as soon as the minimum number of Pods of a Deployment are in a ready state, while KubernetesWrapper will change its status from `Progressing` to `Completed` only after all replicas have been updated.

I tried first to solve this mismatch by using an intermediate `Available`/`Ready` state between `Progressing` and `Completed`, using the Available condition, but that was not a proper solution as I was mixing concerns/conditions.
Then I tried [returning the Available condition in a new property](https://github.com/Altinn/app-kubernetes-wrapper/pull/27/commits/e3948052573de29a9733a1d9aa8585978965b827#diff-b418be66d7bc5ccac3024b1504ea087c205b3a55e7bb69a7bf8f0e18ccb8779aR124), but I encountered issues as it appears that the calculation of the minimum available replicas also includes the old pods 🤷‍♂️ 

> - Since we're using `--atomic`, this can end up in 2 states possibles for the deployment: Healthy or unhealthy. Heallthy are both Completed and Rollbacked. Unhealthy it's only temporary because it'll be rollbacked anyway, right? What I mean is that maybe you don't need to report on Failures because they're temporary — they last as much as the `progressDeadlineSeconds`.

I wanted to report on Failures too as it seems that KubernetesWrapper is not only used to monitor deployments made via the `altinn-studio-deploy-app-image` pipeline, but also other deployments that might not use the `--atomic` flag (e.g. https://platform.tt02.altinn.no/kuberneteswrapper/api/v1/deployments, https://dev.altinn.studio/kuberneteswrapper/api/v1/deployments).

> Good discussion, this isn't a simple problem :) 

Yes, this is more complicated than I thought! 😕 I'm actually wondering if I really still need this PR 🤔 I can now just use the status of the pipeline as we are now using the `--atomic` flag.

**Comment by mlqn** [comment]
I have discussed with my team (team studio). The `--atomic` flag and the version returned by KubernetesWrapper are actually enough for knowing the status of the application (none, in progress, success), so we no longer need this PR.
It can still be relevant to return a status for monitoring services that are not deployed with the `--atomic` flag, but given the complexity, I suggest implementing it when the need arises.
